File: ru/codes/go/chapter_backtracking/subset_sum_i.go
// File: subset_sum_i.go
// Created Time: 2023-06-24
// Author: Reanon (793584285@qq.com)

package chapter_backtracking

import "sort"

/* Алгоритм бэктрекинга: сумма подмножеств I */
func backtrackSubsetSumI(start, target int, state, choices *[]int, res *[][]int) {
	// Если сумма подмножества равна target, записать решение
	if target == 0 {
		newState := append([]int{}, *state...)
		*res = append(*res, newState)
		return
	}
	// Обойти все варианты выбора
	// Отсечение 2: начинать обход с start, чтобы избежать генерации повторяющихся подмножеств
	for i := start; i < len(*choices); i++ {
		// Отсечение 1: если сумма подмножества превышает target, немедленно завершить цикл
		// Это связано с тем, что массив уже отсортирован, следующие элементы больше, и сумма подмножества точно превысит target
		if target-(*choices)[i] < 0 {
			break
		}
		// Попытка: сделать выбор и обновить target и start
		*state = append(*state, (*choices)[i])
		// Перейти к следующему выбору
		backtrackSubsetSumI(i, target-(*choices)[i], state, choices, res)
		// Откат: отменить выбор и восстановить предыдущее состояние
		*state = (*state)[:len(*state)-1]
	}
}

/* Решить задачу суммы подмножеств I */
func subsetSumI(nums []int, target int) [][]int {
	state := make([]int, 0) // Состояние (подмножество)
	sort.Ints(nums)         // Отсортировать nums
	start := 0              // Стартовая вершина обхода
	res := make([][]int, 0) // Список результатов (список подмножеств)
	backtrackSubsetSumI(start, target, &state, &nums, &res)
	return res
}

File: ru/codes/csharp/chapter_backtracking/subset_sum_i.cs
/**
* File: subset_sum_i.cs
* Created Time: 2023-06-25
* Author: hpstory (hpstory1024@163.com)
*/

namespace hello_algo.chapter_backtracking;

public class subset_sum_i {
    /* Алгоритм бэктрекинга: сумма подмножеств I */
    void Backtrack(List<int> state, int target, int[] choices, int start, List<List<int>> res) {
        // Если сумма подмножества равна target, записать решение
        if (target == 0) {
            res.Add(new List<int>(state));
            return;
        }
        // Обойти все варианты выбора
        // Отсечение 2: начинать обход с start, чтобы избежать генерации повторяющихся подмножеств
        for (int i = start; i < choices.Length; i++) {
            // Отсечение 1: если сумма подмножества превышает target, немедленно завершить цикл
            // Это связано с тем, что массив уже отсортирован, следующие элементы больше, и сумма подмножества точно превысит target
            if (target - choices[i] < 0) {
                break;
            }
            // Попытка: сделать выбор и обновить target и start
            state.Add(choices[i]);
            // Перейти к следующему выбору
            Backtrack(state, target - choices[i], choices, i, res);
            // Откат: отменить выбор и восстановить предыдущее состояние
            state.RemoveAt(state.Count - 1);
        }
    }

    /* Решить задачу суммы подмножеств I */
    List<List<int>> SubsetSumI(int[] nums, int target) {
        List<int> state = []; // Состояние (подмножество)
        Array.Sort(nums); // Отсортировать nums
        int start = 0; // Стартовая вершина обхода
        List<List<int>> res = []; // Список результатов (список подмножеств)
        Backtrack(state, target, nums, start, res);
        return res;
    }

    [Test]
    public void Test() {
        int[] nums = [3, 4, 5];
        int target = 9;
        List<List<int>> res = SubsetSumI(nums, target);
        Console.WriteLine("Входной массив nums = " + string.Join(", ", nums) + ", target = " + target);
        Console.WriteLine("Все подмножества с суммой " + target + ": res = ");
        foreach (var subset in res) {
            PrintUtil.PrintList(subset);
        }
    }
}
